man/lz4_decompress.1
'\" t
.\"     Title: lz4_decompress
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 09/08/2025
.\"    Manual: MySQL Database System
.\"    Source: MySQL 8.0
.\"  Language: English
.\"
.TH "LZ4_DECOMPRESS" "1" "09/08/2025" "MySQL 8\&.0" "MySQL Database System"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
lz4_decompress \- decompress mysqlpump LZ4\-compressed output
.SH "SYNOPSIS"
.HP \w'\fBlz4_decompress\ \fR\fB\fIinput_file\fR\fR\fB\ \fR\fB\fIoutput_file\fR\fR\ 'u
\fBlz4_decompress \fR\fB\fIinput_file\fR\fR\fB \fR\fB\fIoutput_file\fR\fR
.SH "DESCRIPTION"
.PP
The
\fBlz4_decompress\fR
utility decompresses
\fBmysqlpump\fR
output that was created using LZ4 compression\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
\fBlz4_decompress\fR
is deprecated as of MySQL 8\&.0\&.34; expect it to be removed in a future version of MySQL\&. This is because the associated
\fBmysqlpump\fR
utility is deprecated as of MySQL 8\&.0\&.34\&.
.sp .5v
.RE
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
If MySQL was configured with the
\fB\-DWITH_LZ4=system\fR
option,
\fBlz4_decompress\fR
is not built\&. In this case, the system
\fBlz4\fR
command can be used instead\&.
.sp .5v
.RE
.PP
Invoke
\fBlz4_decompress\fR
like this:
.sp
.if n \{\
.RS 4
.\}
.nf
lz4_decompress \fIinput_file\fR \fIoutput_file\fR
.fi
.if n \{\
.RE
.\}
.PP
Example:
.sp
.if n \{\
.RS 4
.\}
.nf
mysqlpump \-\-compress\-output=LZ4 > dump\&.lz4
lz4_decompress dump\&.lz4 dump\&.txt
.fi
.if n \{\
.RE
.\}
.PP
To see a help message, invoke
\fBlz4_decompress\fR
with no arguments\&.
.PP
To decompress
\fBmysqlpump\fR
ZLIB\-compressed output, use
\fBzlib_decompress\fR\&. See
zlib_decompress(1)\&.
.SH "COPYRIGHT"
.br
.PP
Copyright \(co 1997, 2025, Oracle and/or its affiliates.
.PP
This documentation is free software; you can redistribute it and/or modify it only under the terms of the GNU General Public License as published by the Free Software Foundation; version 2 of the License.
.PP
This documentation is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along with the program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA or see http://www.gnu.org/licenses/.
.sp
.SH "SEE ALSO"
For more information, please refer to the MySQL Reference Manual,
which may already be installed locally and which is also available
online at http://dev.mysql.com/doc/.
.SH AUTHOR
Oracle Corporation (http://dev.mysql.com/).


utilities/lz4_decompress.cc
/*
   Copyright (c) 2014, 2025, Oracle and/or its affiliates.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is designed to work with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have either included with
   the program or referenced in the documentation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

#include <lz4frame.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "client/client_priv.h"
#include "my_sys.h"
#include "print_version.h"
#include "welcome_copyright_notice.h"

static void usage() {
  print_version();
  puts(ORACLE_WELCOME_COPYRIGHT_NOTICE("2015"));
  puts(
      "Decompress data compressed by mysqlpump using lz4 compression "
      "algorithm by reading from input file and writing uncompressed "
      "data to output file");
  printf("Usage: %s input_file output_file\n", "lz4_decompress");
}

static const int INPUT_BUFFER_SIZE = 1024 * 1024;
static const int OUTPUT_BUFFER_SIZE = 1024 * 1024;

int main(int argc, char **argv) {
  CLIENT_WARN_DEPRECATED_NO_REPLACEMENT("lz4_decompress");
  MY_INIT(argv[0]);
  if (argc != 3) {
    usage();
    exit(1);
  }
  FILE *input_file = fopen(argv[1], "rb");
  FILE *output_file = fopen(argv[2], "wb");
  if (input_file == nullptr) {
    fprintf(stderr,
            "lz4_decompress: [Error] Cannot open input file for reading.\n");
    exit(1);
  }
  if (output_file == nullptr) {
    fprintf(stderr, "lz4_decompress: [Error] Cannot create output file.\n");
    exit(1);
  }
  LZ4F_decompressionContext_t decompression_context = nullptr;
  LZ4F_createDecompressionContext(&decompression_context, LZ4F_VERSION);
  char *input_buffer = new char[INPUT_BUFFER_SIZE];
  char *output_buffer = new char[OUTPUT_BUFFER_SIZE];
  bool is_eof = false;

  while (!is_eof) {
    int bytes_read = fread(input_buffer, 1, INPUT_BUFFER_SIZE, input_file);

    if (bytes_read < INPUT_BUFFER_SIZE) {
      is_eof = feof(input_file);
      if (!is_eof) {
        fprintf(stderr,
                "lz4_decompress: [Error] Encountered problem during reading "
                "input.\n");
        exit(1);
      }
    }

    size_t bytes_processed = 0;

    for (;;) {
      size_t current_bytes_processed = bytes_read - bytes_processed;
      size_t bytes_to_write = OUTPUT_BUFFER_SIZE;
      size_t result = LZ4F_decompress(
          decompression_context, output_buffer, &bytes_to_write,
          input_buffer + bytes_processed, &current_bytes_processed, NULL);
      if (LZ4F_isError(result)) {
        fprintf(stderr,
                "lz4_decompress: [Error] Encountered problem during "
                "decompression.\n");
        exit(1);
      }
      bytes_processed += current_bytes_processed;

      if (bytes_to_write == 0) break;

      if (fwrite(output_buffer, 1, bytes_to_write, output_file) !=
          bytes_to_write) {
        fprintf(stderr,
                "lz4_decompress: [Error] Encountered problem during "
                "writing to buffer.\n");
        exit(1);
      }
    }
  }

  LZ4F_freeDecompressionContext(decompression_context);
  delete[] input_buffer;
  delete[] output_buffer;
  fclose(input_file);
  fclose(output_file);
  my_end(0);
  exit(0);
}
